src/main/java/dataStructures/segmentTree/SegmentTree.java

```diff
@@ -0,0 +1,115 @@
+package dataStructures.segmentTree;
+
+/**
+ * Implementation of a Segment Tree. Uses SegmentTreeNode as a helper node class.
+ */
+public class SegmentTree {
+    private SegmentTreeNode root;
+    private int[] array;
+
+    /**
+     * Helper node class. Used internally.
+     */
+    private class SegmentTreeNode {
+        private SegmentTreeNode leftChild;   // left child
+        private SegmentTreeNode rightChild;  // right child
+        private int start; // start idx of range captured
+        private int end; // end idx of range captured
+        private int sum; // sum of all elements between start and end index inclusive
+
+        /**
+         * Constructor
+         * @param leftChild
+         * @param rightChild
+         * @param start
+         * @param end
+         * @param sum
+         */
+        public SegmentTreeNode(SegmentTreeNode leftChild, SegmentTreeNode rightChild, int start, int end, int sum) {
+            this.leftChild = leftChild;
+            this.rightChild = rightChild;
+            this.start = start;
+            this.end = end;
+            this.sum = sum;
+        }
+    }
+
+    /**
+     * Constructor.
+     * @param nums
+     */
+    public SegmentTree(int[] nums) {
+        root = buildTree(nums, 0, nums.length - 1);
+        array = nums;
+    }
+
+    private SegmentTreeNode buildTree(int[] nums, int start, int end) {
+        if (start == end) {
+            return new SegmentTreeNode(null, null, start, end, nums[start]);
+        }
+        int mid = start + (end - start) / 2;
+        SegmentTreeNode left = buildTree(nums, start, mid);
+        SegmentTreeNode right = buildTree(nums, mid + 1, end);
+        return new SegmentTreeNode(left, right, start, end, left.sum + right.sum);
+    }
+
+    /**
+     * Queries the sum of all values in the specified range.
+     * @param leftEnd
+     * @param rightEnd
+     * @return the sum.
+     */
+    public int query(int leftEnd, int rightEnd) {
+        return query(root, leftEnd, rightEnd);
+    }
+
+    private int query(SegmentTreeNode node, int leftEnd, int rightEnd) {
+        // this is the case when:
+        //                start     end
+        // range query:  ^             ^  --> so simply capture the sum at this node!
```

Comment:
How do you capture the sums that are not part of the interval of the current node?